src/v/cloud_storage/partition_manifest.cc

```diff
@@ -497,8 +499,26 @@ bool partition_manifest::advance_start_offset(model::offset new_start_offset) {
             _start_offset = new_head_segment->second.base_offset;
         }
 
-        for (auto iter = _segments.begin(); iter != new_head_segment; ++iter) {
-            subtract_from_cloud_log_size(iter->second.size_bytes);
+        auto previous_head_segment = segment_containing(previous_start_offset);
+        if (previous_head_segment == end()) {
+            // This branch should never be taken. It indicates that the
+            // in-memory manifest may be is in some sort of inconsistent state.
+            vlog(
+              cst_log.warn,
+              "Previous start offset is not within segment in "
+              "manifest for {}: previous_start_offset={}",
+              _ntp,
+              previous_start_offset);
+            previous_head_segment = _segments.begin();
+        }
+
+        // Note that we start subtracting from the cloud log size from the
+        // previous head segments. This is required in order to account for
+        // the case when two `truncate` commands are applied sequentially,
+        // without a `cleanup_metadata` command in between to trim the list of
+        // segments.
+        for (auto it = previous_head_segment; it != new_head_segment; ++it) {
```

Comment:
it's hard to "see" how the property of `previous_head_segment = segment_containing(previous_start_offset);` being sequenced before (or being equal to) `new_head_segment` in `_segments` is maintained. and when it isn't, it looks like undefined behavior. how can we make that clearer or maybe we need and assertion that this iteration doesn't fall off the end?

Reply:
Good point. The manifest would have to be in an inconsistent state for that to happen, but we should guard against it. I'll update the PR.

Reply:
These iterators point to an ordered map of segments that are key-ed by their start offset. This means that as long as the previous start offset is smaller or equal to the new start offset this iteration is fine. I've added an early error exit path from the function if the new start offset is smaller than the previous one.